pisa/stages/osc/pi_2nu.py

```diff
@@ -0,0 +1,147 @@
+
+    @profile
+    def apply_function(self):
+
+        theta23 = self.params.theta23.value.m_as('rad')
+        deltam31 = self.params.deltam31.value.m_as('eV**2')
+
+        for container in self.data:
+            if 'numu' in container.name:
+              apply_probs_vectorized(container['nu_flux'].get(WHERE),
+                                    FTYPE(theta23),
+                                    FTYPE(deltam31),
+                                    container['true_energy'].get(WHERE),
+                                    container['true_coszen'].get(WHERE),
+                                    ITYPE(1),
+                                    out=container['weights'].get(WHERE),
+                                   )
+
+            if 'nutau' in container.name:
+              apply_probs_vectorized(container['nu_flux'].get(WHERE),
+                                    FTYPE(theta23),
+                                    FTYPE(deltam31),
+                                    container['true_energy'].get(WHERE),
+                                    container['true_coszen'].get(WHERE),
+                                    ITYPE(3),
+                                    out=container['weights'].get(WHERE),
+                                   )
+            if 'nue' in container.name:
+              apply_probs_vectorized(container['nu_flux'].get(WHERE),
+                                    FTYPE(theta23),
+                                    FTYPE(deltam31),
+                                    container['true_energy'].get(WHERE),
+                                    container['true_coszen'].get(WHERE),
+                                    ITYPE(0),
+                                    out=container['weights'].get(WHERE),
+                                   )            
+            container['weights'].mark_changed(WHERE) 
+
+@myjit
+def calc_probs(t23, dm31, true_energy, true_coszen): #
+    ''' calculate osc prob of numu to nutau '''
+    L1 = 19. # atmospheric production height 
+    R = 6378.2 + L1 # mean radius of the Earth + L1
+    phi = np.arcsin((1-L1/R)*np.sin(np.arccos(true_coszen)))
+    psi = np.arccos(true_coszen) - phi
+    propdist = np.sqrt( (R-L1)**2 + R**2 - (2*(R-L1)*R*np.cos(psi)))
+
+    return t23*np.sin(1.267*dm31*propdist/true_energy)**2 # constant arise from conversion of units for L and E
+
+# vectorized function to apply (flux * prob)
+# must be outside class
+if FTYPE == np.float64:
+    FX = 'f8'
+    IX = 'i8'
+    signature = '(f8[:], f8, f8, f8, f8, i4, f8[:])'
+else:
+    FX = 'f4'
+    IX = 'i4'
+signature = f'({FX}[:], {FX}, {FX}, {FX}, {FX}, {IX}, {FX}[:])'
+@guvectorize([signature], '(d),(),(),(),(),()->()', target=TARGET)
+def apply_probs_vectorized(flux, t23, dm31, true_energy, true_coszen, nuflav, out):
+    if nuflav==1: # numu receive weights dependent on numu survival prob
+        out[0] *= flux[0] + (flux[1] * (1.0-calc_probs(t23, dm31, true_energy, true_coszen)))
```

Comment:
Just realized here: why do you add the electron flux for both, numu and nutau?
Wouldn't you want just `flux[1] * calc_probs(...)` ?
Looks like a bug...
Did you validate the stage outputs and compare for example with 3 flavour oscillations?